src/cat.jl

```diff
@@ -34,56 +12,57 @@ for (T, S) in [
-        N = length(newL)
-        return NamedDimsArray{newL, T, N, Array{T,N}}(data)
+        data = Base.cat(unname(a), unname(b), unname.(cs)...; dims=numerical_dims)
+        return NamedDimsArray{newL}(data)
     end
 end
 
-# vcat and hcat
 function Base.hcat(a::NamedDimsArray{L}) where L
     newL = expand_dimnames(L, 2)
     data = Base.hcat(parent(a))
-    T = eltype(a)
-    N = length(newL)
-    return NamedDimsArray{newL, T, N, Array{T,N}}(data)
+    return NamedDimsArray{newL}(data)
 end
 
 Base.vcat(a::NamedDimsArray{L}) where L = a
 
-# Base.hcat and Base.vcat specialise on this Union
-const AbsVecOrMat = Union{AbstractVector, AbstractMatrix}
 for (T, S) in [
-    (:NamedDimsArray, :AbsVecOrMat),
-    (:NamedDimsArray, :AbstractArray),
-    (:AbsVecOrMat, :NamedDimsArray),
-    (:AbstractArray, :NamedDimsArray),
-    (:NamedDimsArray, :NamedDimsArray)
+    (:NamedDimsVecOrMat, :NamedDimsVecOrMat),
+    (:NamedDimsVecOrMat, :AbstractVecOrMat),
+    (:AbstractVecOrMat, :NamedDimsVecOrMat),
     ]
-
     for (fun, d) in zip((:vcat, :hcat), (1, 2))
-        # this special case is needed to resolve ambiguities
-        @eval function Base.$fun(a::$T, b::$S)
-            combL = unify_names_shortest(dimnames(a), dimnames(b))
+
+        @eval function Base.$fun(a::$T, b::$S, cs::AbstractVecOrMat...)
+            combL = unify_names_longest(dimnames(a), dimnames(b), dimnames.(cs)...)
             newL = expand_dimnames(combL, $d)
-            data = Base.$fun(parent(a), parent(b))
-            T = promote_type(eltype(a), eltype(b))
-            N = length(newL)
-            return NamedDimsArray{newL, T, N, Array{T,N}}(data)
+            data = Base.$fun(unname(a), unname(b), unname.(cs)...)
+            return NamedDimsArray{newL}(data)
         end
+    end
+end
 
-        @eval function Base.$fun(a::$T, b::$S, c::NamedDimsArray...)
-            combL = unify_names_shortest(dimnames(a), dimnames(b), ntuple(i->dimnames(c[i]), length(c))...)
-            newL = expand_dimnames(combL, $d)
-            data = Base.$fun(parent(a), parent(b), ntuple(i->parent(c[i]), length(c))...)
-            T = promote_type(eltype(a), eltype(b), ntuple(i->eltype(c[i]), length(c))...)
-            N = length(newL)
-            return NamedDimsArray{newL, T, N, Array{T,N}}(data)
+for (f, nf, tf, tup) in [
+    (:vcat, :_named_vcat, :_typed_vcat, ()),
+    (:hcat, :_named_hcat, :_typed_hcat, (:_,)),
+    ]
+    @eval begin
+        Base.reduce(::typeof($f), A::AbstractVector{<:NamedDimsVecOrMat}) =
+            $nf(mapreduce(dimnames, unify_names_longest, A), A)
+        Base.reduce(::typeof($f), A::NamedDimsVector{<:Any,<:AbstractVecOrMat}) =
+            $nf(mapreduce(dimnames, unify_names_longest, A), A)
+        Base.reduce(::typeof($f), A::NamedDimsVector{<:Any,<:NamedDimsVecOrMat}) =
+            $nf(mapreduce(dimnames, unify_names_longest, A), A)
+
+        function $nf(Linner, A)
+            Louter = ($tup..., dimnames(A)...)
```

Comment:
Why does `hcat` get an extra unnamed dimension?

Reply:
Because the reduction is always along 2nd dimension, so the outer container’s name, if any, must not be first.

Reply:
ok, lgtm

Reply:
Cool. I don't actually have a working merge button on this package, but maybe you do?